source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,863 @@
+
+Backwards Breaking Considerations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This specification deprecates many existing timeout options and introduces a
+new exception type that is used to communicate timeout expiration. If drivers
+need to make backwards-breaking changes to support ``timeoutMS``, the
+backwards breaking behavior MUST be gated behind the presence of the
+``timeoutMS`` option. If the ``timeoutMS`` option is not set, drivers MUST
+continue to honor existing timeouts such as ``socketTimeoutMS``. Backwards
+breaking changes include any changes to exception types thrown by stable API
+methods or changes to timeout behavior. Drivers MUST document these changes.
+
+In a subsequent major release, drivers SHOULD drop support for legacy timeout
+behavior and only continue to support the timeout options that are not
+deprecated by this specification. Once legacy options are removed, drivers
+MUST make the backwards-breaking behavioral changes described in this
+specification regardless of whether or not ``timeoutMS`` is set by the
+application.
+
+See the `Errors`_ section for explanations of the backwards-breaking changes
+to error reporting.
+
+Deprecations
+~~~~~~~~~~~~
+
+The following configuration timeout options MUST be deprecated in favor of
+``timeoutMS``:
+
+- ``socketTimeoutMS``
+
+- ``waitQueueTimeoutMS``
+
+- ``wTimeoutMS``
+
+The following options for CRUD methods MUST be deprecated in favor of
+``timeoutMS``:
+
+- ``maxTimeMS``
+
+- ``maxCommitTimeMS``
+
+Drivers MUST throw an error during MongoClient configuration if ``timeoutMS``
+is used in conjunction with one of the options in the deprecated list. This
+validation MUST NOT be done during MongoDatabase or MongoCollection
+configuration.
+
+Timeout Behavior
+----------------
+
+The ``timeoutMS`` option specifies the best-effort maximum amount of time a
+single operation can take before control is returned to the application.
+Drivers MUST keep track of the remaining time before the timeout expires as
+an operation progresses.
+
+Operations
+~~~~~~~~~~
+
+The ``timeoutMS`` option applies to all operations defined in the following
+specifications:
+
+- `CRUD <./../crud/crud.rst>`__
+- `Change Streams <../change-streams/change-streams.rst>`__
+- `Client Side Encryption <../client-side-encryption/client-side-encryption.rst>`__
+- `Enumerating Collections <../enumerate-collections.rst>`__
+- `Enumerating Databases <../enumerate-databases.rst>`__
+- `GridFS <../gridfs/gridfs-spec.rst>`__
+- `Index Management <../index-management.rst>`__
+- `Transactions <../transactions/transactions.rst>`__
+
+In addition, it applies to all operations on cursor objects that may perform
+blocking work (e.g. methods to iterate or close a cursor, any method that
+reads documents from a cursor into an array, etc).
+
+Validation and Overrides
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+When executing an operation that supports the ``timeoutMS`` option, drivers
```

Comment:
Added links to the same two rationale sections that I mentioned in your previous comments.

Regarding your example: do you have any ideas for improving it? I agree it's a valid concern, but I feel like it's already a problem for existing options. For example, there's a similar potential problem if an application is tested with one set of TLS certs and everything works but the production MongoClient is configured with different certificates which are expired. This case is arguably worse because all operations would have for ~30 seconds and fail with server selection timeout errors.

Reply:
For an example more similar to this one, consider the case of a DAO executing an operation with an explicit ClientSession and the production MongoClient being configured with an unacknowledged write concern. Per the [sessions spec](https://github.com/mongodb/specifications/blob/8c67a52640cba3563897a1d1c8ef30a3878b7054/source/sessions/driver-sessions.rst#when-using-unacknowledged-writes), this would raise a client-side error.

Reply:
> do you have any ideas for improving it?

I proposed an alternative in the new rationale section.

I'm not persuaded by your counter-examples.   The one about TLS is quite different, because it's just broken no matter what a developer does in their DAO.  The one about `w:0` writes is more on point, but in that case we're talking about a global option that is exceedingly rare to set as `w:0`.  We would have made a different choice it it was an error on a more common value of write concern.